File: TestLib/src/main/java/com/example/HashTable/VerifyinganAlienDictionary_953.java
package com.example.HashTable;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

//Input: words = ["hello","leetcode"], order = "hlabcdefgijkmnopqrstuvwxyz"
//        Output: true
//        Explanation: As 'h' comes before 'l' in this language, then the sequence is sorted.
//        https://leetcode.com/problems/verifying-an-alien-dictionary/

public class VerifyinganAlienDictionary_953 {

    public static List<List<Integer>> levelOrder = new ArrayList<List<Integer>>();
    public static void main(String[] args) {
        String[] words = {"hello", "leetcode"};
        System.out.print(isAlienSorted(words, "hlabcdefgijkmnopqrstuvwxyz"));
    }

    public static boolean isAlienSorted(String[] words, String order) {
        int[] char_map  = new int[26];
        for (int i =0; i < order.length(); i++) {
            char_map[order.charAt(i)-'a'] =  i;
        }
        for (int i=1; i < words.length;i++) {
            if (compare(words[i-1], words[i], char_map) > 0) {
              return false;
            }
        }
        return true;
    }

    public static int compare(String word1, String word2, int[] char_map) {
        int char_result =0;
        int i =0, j = 0;
        while(i < word1.length() && j < word2.length() && char_result == 0) {
            char_result = char_map[word1.charAt(i) - 'a'] - char_map[word2.charAt(j) - 'a'];
            i++;
            j++;
        }

        if (char_result == 0) {
            return word1.length()-word2.length();
        } else {
            return char_result;
        }
    }
}

File: TestLib/src/main/java/com/example/Kotlin/VerifyinganAlienDictionary_953.kt
package com.example.Kotlin

//Input: words = ["hello","leetcode"], order = "hlabcdefgijkmnopqrstuvwxyz"
//        Output: true
//        Explanation: As 'h' comes before 'l' in this language, then the sequence is sorted.
//        https://leetcode.com/problems/verifying-an-alien-dictionary/

fun main (args: Array<String>) {
    println(isAlienSorted(arrayOf("hello","leetcode"), "hlabcdefgijkmnopqrstuvwxyz"))
}

fun isAlienSorted(words: Array<String>, order: String): Boolean {
    var charMap =  IntArray(26)
    for (i in 0 until order.length) {
        charMap[order[i]-'a'] = i    //set the order based on the position
    }
    for (i in 1 until words.size) {
        if (compare(words[i-1], words[i], charMap) > 0) { //has to be less than zero
            return false;
        }
    }
    return true
}

fun compare(word1: String, word2: String, charMap: IntArray) : Int {
    var currentcount = 0
    var i=0; var j=0
    while(i < word1.length && j < word2.length && currentcount ==0) {
        currentcount +=  charMap[word1[i]-'a'] - charMap[word2[j]-'a']
        i++
        j++
    }
    if (currentcount==0) {
        return word1.length-word2.length
    }
    return currentcount
}